File: January2022/Python/LinkedListRandomNode.py
#Author :  Karthik K (https://www.linkedin.com/in/karthikmunirathinam/)

'''
        Approach 1  
        TC : O(1)
        SC : O(n)
'''

class Solution:
    import random
    def __init__(self, head: Optional[ListNode]):
        self.array=[]
        while head:
            self.array.append(head.val)
            head=head.next
        

    def getRandom(self) -> int:
        return random.choice(self.array)
       
'''
####### Approach 2 ###### 
        TC : O(n)
        SC : O(1)
'''
class Solution:
    def __init__(self, head: Optional[ListNode]):
        self.head=head
        
    def getRandom(self) -> int:
        reservoir=self.head.val
        i=1
        curr=self.head
        while curr:
            if random.random()<1/i:
                reservoir=curr.val
            i+=1
            curr=curr.next
        return reservoir
        

File: January2022/C++/Linked_List_Random_Node.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* head;
    public:
    Solution(ListNode* head) {
        this->head=head;
    }
    
    int getRandom() {
        int res=this->head->val;
        ListNode* node=this->head->next;
        int i=2;
        while(node)
        {
            int r=rand()%i;
            if(r==0)
                res=node->val;
            i++;
            node=node->next;
        }
        return res;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(head);
 * int param_1 = obj->getRandom();
 */